Mathlib/Data/Stream/Init.lean

```diff
@@ -0,0 +1,795 @@
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+theorem corec_eq (f : α → β) (g : α → α) (a : α) : corec f g a = f a::corec f g (g a) := by
+  rw [corec_def, map_eq, head_iterate, tail_iterate]; rfl
+#align stream.corec_eq Stream'.corec_eq
+
+theorem corec_id_id_eq_const (a : α) : corec id id a = const a := by
+  rw [corec_def, map_id, iterate_id]
+#align stream.corec_id_id_eq_const Stream'.corec_id_id_eq_const
+
+theorem corec_id_f_eq_iterate (f : α → α) (a : α) : corec id f a = iterate f a :=
+  rfl
+#align stream.corec_id_f_eq_iterate Stream'.corec_id_f_eq_iterate
+
+end Corec
+
+section Corec'
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+theorem corec'_eq (f : α → β × α) (a : α) : corec' f a = (f a).1::corec' f (f a).2 :=
+  corec_eq _ _ _
+#align stream.corec'_eq Stream'.corec'_eq
+
+end Corec'
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+theorem unfolds_eq (g : α → β) (f : α → α) (a : α) : unfolds g f a = g a::unfolds g f (f a) := by
+  unfold unfolds; rw [corec_eq]
+#align stream.unfolds_eq Stream'.unfolds_eq
+
+theorem nth_unfolds_head_tail : ∀ (n : Nat) (s : Stream' α),
+    nth (unfolds head tail s) n = nth s n :=
+  by
+  intro n; induction' n with n' ih
+  · intro s
+    rfl
+  · intro s
+    rw [nth_succ, nth_succ, unfolds_eq, tail_cons, ih]
+#align stream.nth_unfolds_head_tail Stream'.nth_unfolds_head_tail
+
+theorem unfolds_head_eq : ∀ s : Stream' α, unfolds head tail s = s := fun s =>
+  Stream'.ext fun n => nth_unfolds_head_tail n s
+#align stream.unfolds_head_eq Stream'.unfolds_head_eq
+
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+/- ./././Mathport/Syntax/Translate/Expr.lean:177:8: unsupported: ambiguous notation -/
+theorem interleave_eq (s₁ s₂ : Stream' α) : s₁ ⋈ s₂ = head s₁::head s₂::(tail s₁ ⋈ tail s₂) := by
+  let t := tail s₁ ⋈ tail s₂
+  show s₁ ⋈ s₂ = head s₁::head s₂::t
+  unfold interleave; unfold corecOn; rw [corec_eq]; dsimp; rw [corec_eq]; rfl
+#align stream.interleave_eq Stream'.interleave_eq
+
+theorem tail_interleave (s₁ s₂ : Stream' α) : tail (s₁ ⋈ s₂) = s₂ ⋈ tail s₁ := by
+  unfold interleave corecOn; rw [corec_eq]; rfl
+#align stream.tail_interleave Stream'.tail_interleave
+
+theorem interleave_tail_tail (s₁ s₂ : Stream' α) : tail s₁ ⋈ tail s₂ = tail (tail (s₁ ⋈ s₂)) := by
+  rw [interleave_eq s₁ s₂]; rfl
+#align stream.interleave_tail_tail Stream'.interleave_tail_tail
+
+theorem nth_interleave_left : ∀ (n : Nat) (s₁ s₂ : Stream' α),
+    nth (s₁ ⋈ s₂) (2 * n) = nth s₁ n := by
+  intro n s₁ s₂
+  match n, s₁, s₂ with
+  |0, s₁, s₂ => rfl
+  |n + 1, s₁, s₂ =>
+    change nth (s₁ ⋈ s₂) (succ (succ (2 * n))) = nth s₁ (succ n)
+    rw [nth_succ, nth_succ, interleave_eq, tail_cons, tail_cons]
+    have : n < succ n := Nat.lt_succ_self n
+    let ih := nth_interleave_left n (tail s₁) (tail s₂)
+    rw [ih]
+    rfl
```

Comment:
You don't need the `have` and you can just add `nth_interleave_left n (tail s₁) (tail s₂)` to the `rw`

Reply:
Done. Thanks.